### PR TITLE
add persistable_segment_tree.md

### PR DESCRIPTION
**Pull request by valderyaya**
- [x] Read [Contribute to NTHU-CPP](https://nthu-cp.github.io/NTHU-CPP/others/contribution.html).
- [x] Rebase to the latest main branch.
- [x] List all the references.
- [x] Successfully build the website by mdBook with no error after the change.
- [x] Exclude any irrelevant files.
- [x] Link to the issue which is related to your change.
- [x] Review the content by myself.
- [x] Pass tests/CI.

### CODE REVIEW

**Comment by valderyaya** [comment]
latex語法待改

**Comment by valderyaya** [comment]
Finish

**Comment by valderyaya** [comment]
@harry900831  Waiting review

File: src/data_structures/persistable_segment_tree.md [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+<img src="image/persistable_segment_tree/persistable_segment_tree-2.png">
+
+## 回到題目
+
+現在回到題目上，如果我們把區間這個條件修改一下，每次詢問 \\([1, r]\\) 的第 \\(k\\) 小，我們就只要找到第\\(r\\) 個版本，利用權值線段樹求出來就可以了
+
+
+那對於區間 \\([l,r]\\)，我們可以利用前綴和的特性，用第 \\(r\\) 個版本的線段樹減掉第 \\(l-1\\) 個版本就可以求出 \\([l,r]\\) 這個區間的權值線段樹的樣子，這樣答案就出來了。
+
+有一點要注意的是，持久化線段樹需要使用動態開點來實作，如果使用僞指標的話就要特別注意一下記憶體的大小，如果有 \\(n\\) 次修改，那每次修改需要增加 \\(\lceil log_2{n} \rceil + 1\\) 個點，線段樹總空間會到 \\(2n -1 + n(\lceil log_2{n} \rceil + 1)\\)。 依照題目大小帶入數字 \\(n \leq 2\times10^5\\)， \\(2 \times 2\times10^5 -1 + 2\times10^5 \times 19\\) 約等於 \\(21 \times 2\times10^5\\)，因此空間需要開到 \\(21\\) 倍，而因爲題目一般情況下不會特別限制空間大小，所以爲了方便都會寫成 `n<<5`，即開到32倍。
```
Comment (harry900831) [inline src/data_structures/persistable_segment_tree.md:53]: 你需要更詳細的解釋時間與空間複雜度

File: src/data_structures/persistable_segment_tree.md [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+
+附上 [洛谷 P3834](https://www.luogu.com.cn/problem/P3834) 的code
```
Comment (harry900831) [inline src/data_structures/persistable_segment_tree.md:55]: Please use the online judge that we usually use, such as CSES, atcoder, CF.

Here's a problem for you : https://cses.fi/problemset/task/1737

File: src/data_structures/persistable_segment_tree.md [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+    build(1, len, root[0]);
+    for(int i = 1;i <= n; ++i)
+        add(1, len, root[i], root[i-1], lower_bound(b+1,b+1+len,a[i]) - b);
+    for(int l, r, k; m--;){
+        cin >> l >> r >> k;
+        cout << b[query(root[l-1],root[r],1,len,k)] << "\n";
+    }
+    return 0;
+}
+```
```
Comment (harry900831) [inline src/data_structures/persistable_segment_tree.md:106]: Please wrap the segment tree as a struct

File: src/data_structures/persistable_segment_tree.md [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+
+
+給定一個長度為 \\(N\\) 的整數序列 \\(B\\)。有 \\(M\\) 筆詢問，對於每組詢問 \\((P, K)\\)，求最小的答案 \\(S\\) 滿足至少有 \\(K\\) 個元素滿足以下條件
+
+- \\(|P - i| \leq S\\)
+- \\(B_i \leq S\\)
+
+\\(1 \leq N, M \leq 10^5 \\)
+\\(1 \leq K, B_i \leq N \\)
+\\(0 \leq P \lt N\\)
```
Comment (harry900831) [inline src/data_structures/persistable_segment_tree.md:122]: Please follow the format:
https://nthu-cp.github.io/NTHU-CPP/sqrt/sqrt_decomposition.html

File: src/data_structures/persistable_segment_tree.md [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+#define REP1(i,a,b) for(register int i=(a);i<=(b);++i)
+#define em emplace_back
+#define mkp make_pair
+#define pb push_back
+#define ALL(x) (x).begin(),(x).end()
+#define pii pair<int,int>
+#define pll pair<ll,ll>
+#define Rosario ios::sync_with_stdio(0),cin.tie(0),cout.tie(0);
+//#define lb(x) (x&-x)
+
```
Comment (harry900831) [inline src/data_structures/persistable_segment_tree.md:148]: Why the coding style are different in the whole article?

Please use your own code....

Comment (harry900831) [reply src/data_structures/persistable_segment_tree.md:148]: Please remove the useless Marco and your personal habit on the code....

File: src/data_structures/persistable_segment_tree.md (hunk 2 above)
Comment (harry900831) [reply src/data_structures/persistable_segment_tree.md:55]: 對於真的找不到好 OJ但是又是很經典的題目 例如區間地k小 才考慮使用冷門OJ

File: src/data_structures/persistable_segment_tree.md [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+```
+
+### [洛谷 P2633 Count on a tree](https://www.luogu.com.cn/problem/P2633)
+
+#### 題目敘述
+給定一棵 \\(n\\) 個節點的樹，每個點有一個權值。有 \\(m\\) 個詢問，每次給你 \\(u,v,k\\)，你需要回答 \\(u \text{ xor last}\\) 和 \\(v\\) 這兩個節點間第 \\(k\\) 小的點權。  
+
+其中 \\(\text{last}\\) 是上一個詢問的答案，定義其初始為 \\(0\\)。
+
+\\(1\le n,m \le 10^5\\)， 點權在 \\([1,2^{31}-1]\\) 之間。
```
Comment (harry900831) [inline src/data_structures/persistable_segment_tree.md:295]: You use a lot of problem from 洛谷 but I believe there should also have lots of problem that can be solved by persistent segment tree on codeforces or atcoder

File: src/data_structures/persistable_segment_tree.md [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+
+https://tioj.ck.tp.edu.tw/problems/1827
+
+https://www.luogu.com.cn/problem/P2617
+
+https://www.luogu.com.cn/problem/P2633
+
+https://www.luogu.com.cn/problem/P3168
+
+https://codeforces.com/problemset/problem/1771/F
```
Comment (harry900831) [inline src/data_structures/persistable_segment_tree.md:550]: Please use bullet list, problem shouldn't be list on the references.

File: src/data_structures/persistable_segment_tree.md [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+}
+```
+
+
+
+
+## References
+https://blog.csdn.net/yanweiqi1754989931/article/details/117380913
+
+https://oi-wiki.org/ds/persistent-seg/
```
Comment (harry900831) [inline src/data_structures/persistable_segment_tree.md:540]: If you only read the above two resources, it's pretty obvious that you have to do more research...

- https://usaco.guide/adv/persistent?lang=cpp
- https://cp-algorithms.com/data_structures/segment_tree.html#preserving-the-history-of-its-values-persistent-segment-tree
- https://codeforces.com/blog/entry/49777
- https://codeforces.com/blog/entry/56880

File: src/data_structures/persistable_segment_tree.md (hunk 2 above)
Comment (harry900831) [reply src/data_structures/persistable_segment_tree.md:55]: In my opinion, https://www.spoj.com/problems/MKTHNUM/ SPOJ is better than 落谷

**Review by harry900831 — CHANGES_REQUESTED** [review]
- There's a lot of space for improvement on this article.
- Please read more online resources instead of only reading something like OI-wiki, CSDN blog....
- The explanation is not clear, the code need to refactor, the format is incorrect, .... Please re-organize the whole article
- You may refer to others PR or the article that has been merged into NTHU-CPP to know that what kind of article you have to write.
- There's some application for persistent segment tree, If possible, you should also mention it. For example: https://codeforces.com/blog/entry/64628
- Please fix markdownlint and textlint.
- **You should better put more effort on this** otherwise it might not be merged into NTHU-CPP in the summer vacation.

**Comment by harry900831** [comment]
why are you closing this?

@valderyaya

**Comment by valderyaya** [comment]
> why are you closing this?
> 
> @valderyaya 

我在修改的時候可能不小心按到了什麽，我再試著改回來

**Comment by valderyaya** [comment]
@harry900831 我在pull rebase的時候遇到了一些問題，所以我把branch刪掉重建了，這導致了pull request被close，好像沒辦法restore，我可以修改完之後再新開一個pull request嗎

**Pull request closed without merging**